[Power Support] Enable weaveworks scope on Power.

meghalidhoble opened the pull request:
Observed this particular test  was failing on ppc64le architecture. Added the test failure log below, 
  walker_linux_test.go:97:
```                --- want
                +++ have
                @@ -32,7 +32,7 @@
                   Cmdline: (string) (len=15) "curl google.com",
                   Threads: (int) 1,
                   Jiffies: (uint64) 0,
                -  RSSBytes: (uint64) 8192,
                +  RSSBytes: (uint64) 131072,
                   RSSBytesLimit: (uint64) 2048,
                   OpenFilesCount: (int) 3,
                   OpenFilesLimit: (uint64) 32768,
```
Hence modified the value for RSSBytes to accept variable value which is computed based on pageSize instead of hard-coded values, as suggested by Adam Harrison over slack channel. 

I have verified this patch on x86 setup as well and all the tests pass there.  Please provide your comments/suggestions on this. 
Thanks,
Meghali

bboreham commented:
Note there is a lint failure in this PR.

meghalidhoble commented:
Yes, @bboreham , I am looking into this. I am expecting my build-changes should get this resolved.

satyamz commented:
@meghalidhoble Please run `make lint` locally to verify linting issues.

satyamz commented on client/Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM node:8.4.0
+FROM node:8.11
```
Why are you upgrading versions of most of the components in `Dockerfile` ? Why it is required?

meghalidhoble replied:
I have upgraded the versions, as these versions are now mulit-arch supported and will help building scope on many platforms easily. For me it helped to build on ppc64le.

satyamz replied:
ok!  We need to make sure that these upgrades don't break anything.

satyamz commented on probe/endpoint/dns_snooper_linux_ppc64le.go:
```diff
@@ -0,0 +1,284 @@
+package endpoint
```
Why this file is required? This file is exact copy of `github.com/weaveworks/scope/probe/endpoint/dns_snooper_linux_amd64.go` Please explain this in more details.

meghalidhoble replied:
@satyamz , that's right. The new file is exact copy of dns_snooper_linux_amd64.go. I had to add this file to fix the build error which says "undefined: DNSSnooper". Also observed that there is already a file dns_snooper_others.go, however the implementation was missing for the class "DNSSnooper", hence thought of reusing the amd64 file. Please suggest if this can be implemented in different way.

satyamz replied:
Can I have build failure log?

bboreham replied:
The reason it was undefined is the `others` file has a build constraint saying it only builds on `darwin arm`.
Given the stated reason for this PR, the simple way forward would be to add your architecture to that list.

If you can test that the amd64 file works correctly on ppc64le then renaming it and adding a build constraint for those two platforms would be fine.

Copying the whole file is not ok.

meghalidhoble replied:
Hi @bboreham , @satyamz ,
I could add the build-constraint instead of having new file. For this I had to rename the existing file "dns_snooper_linux_amd64.go" to "dns_snooper.go" as the build-tags won't work with file-suffixes. 
Then added a line at the start of the file "dns_snooper.go" as below: 
  ```// +build linux,amd64 ppc64le ``` 

I have verified this patch works on ppc as well as on x86.
Please suggest if this approach looks good, I will commit the changes accordingly.

bboreham commented:
Please run `gofmt -s -w probe/process/walker_linux_test.go` then update the PR.  This should clear the lint error.

meghalidhoble commented:
Thanks @bboreham , your suggestion helped to resolve the failure.

bboreham commented on backend/Dockerfile:
```diff
@@ -1,14 +1,32 @@
+	go install -race -tags netgo std; \
+   fi
+RUN export arch_val="$(dpkg --print-architecture)"; \
+   if [ "$arch_val" != "ppc64el" ]; then \
```
surely if the thing we are trying to download is `amd64` this test should be `== amd64` ?

bboreham commented on backend/Dockerfile:
```diff
@@ -1,14 +1,32 @@
-FROM golang:1.10.2-stretch
+FROM golang:1.10.2
```
please explain this change

meghalidhoble replied:
"golang:1.10.2" is a multi-arch image, which means it will build for amd64 & ppc64le as well. Hence the change. Sorry for not adding the description at the time of commit. I will make sure I will do that now on wards.

bboreham commented:
Please put into each commit message a description of the changes.  
Some tips [here](https://github.com/weaveworks/scope/blob/master/CONTRIBUTING.md#format-of-the-commit-message)
"code changes made" is not a useful description.

abdasgupta commented:
Hi @bboreham, @satyamz ,
Can we please merge this PR if no more modification is needed?

meghalidhoble commented on backend/Dockerfile:
```diff
@@ -1,14 +1,32 @@
-    chmod +x shfmt && \
-    mv shfmt /usr/bin
+   export arch_val="$(dpkg --print-architecture)"; \
+   if [ "$arch_val" != "ppc64el" ]; then \
```
I had to exclude as the race-detector libs are not supported for power.

meghalidhoble commented on probe/endpoint/dns_snooper.go:
```diff
@@ -1,3 +1,7 @@
+// +build linux,amd64 ppc64le
```
The existing file name indicated Linux and amd64, hence the build constraint also indicates the same. However for ppc64le , I want "Linux" to be the OS. I miss-read the separators. Sorry for that, I will correct it.

meghalidhoble commented on probe/endpoint/dns_snooper.go:
```diff
@@ -1,4 +1,5 @@
-// +build linux,amd64 ppc64le
+// +build linux
+// +build amd64 ppc64le
```
Not sure why this change is causing a test-failure, as the command "COVERDIR=./coverage make RM= CODECGEN_UID=23 tests" working fine for me on my setup. 
Any pointers would really help.

meghalidhoble replied:
Could resolve this failure, by latest commit.

meghalidhoble commented:
Hi @bboreham ,
I have addressed most of your comments and have pushed a few new code changes, Could you please take a look and let me know if any comments/suggestions?

meghalidhoble commented:
Hi @bboreham , squashed all the commits to a single commit, with all the implementation details for each file being modified. Please take a look.

bboreham commented:
Sadly, it is inappropriate to directly modify anything in the `tools` directory because it is a vendored copy of https://github.com/weaveworks/build-tools.  Need to modify it upstream and then bring that commit in.

I'm sure this is maddening.

I might just change it to do nothing if it can't find `shfmt`, as is done for `shellcheck`, rather than have an architecture-specific test.

meghalidhoble commented:
Hi @bboreham ,
I have addressed the concern raised for tools directory changes, hence I have removed the architecture specific changes. Sorry for multiple commits, I will squash those commits and will combine to a single commit. Let me know if these changes look ok.

bboreham commented:
Multiple commits is fine as long as they all form a progression and are all self-explanatory.

However the issue is that to change a file under the `tools` directory in this repo you need to change it in https://github.com/weaveworks/build-tools first and then do a `git subtree` to update the directory in this repo, as described in `tools/README.md`.

meghalidhoble commented:
Hi @bboreham , I have raised a pull request at build-tools repo with the lint changes, once those changes are approved, I will run git subtree to update the directory in this repo.

bboreham commented:
Thanks!
